pkg/audit/workflow_auditor.go

```diff
@@ -0,0 +1,410 @@
+		TargetKeyStepType:   stepType,
+		TargetKeyType:       TargetTypeWorkflowStep,
+	}
+	if toolName != "" {
+		target[TargetKeyToolName] = toolName
+	}
+	event.WithTarget(target)
+
+	event.LogTo(ctx, w.auditLogger, LevelAudit)
+}
+
+// LogStepCompleted logs successful step completion.
+func (w *WorkflowAuditor) LogStepCompleted(
+	ctx context.Context,
+	workflowID string,
+	stepID string,
+	duration time.Duration,
+	retryCount int,
+) {
+	if !w.config.ShouldAuditEvent(EventTypeWorkflowStepCompleted) {
+		return
+	}
+
+	source := w.extractSource(ctx)
+	subjects := w.extractSubjects(ctx)
+
+	event := NewAuditEvent(
+		EventTypeWorkflowStepCompleted,
+		source,
+		OutcomeSuccess,
+		subjects,
+		w.component,
+	)
+
+	target := map[string]string{
+		TargetKeyWorkflowID: workflowID,
+		TargetKeyStepID:     stepID,
+		TargetKeyType:       TargetTypeWorkflowStep,
+	}
+	event.WithTarget(target)
+
+	event.Metadata.Extra = map[string]any{
+		MetadataExtraKeyDuration:   duration.Milliseconds(),
+		MetadataExtraKeyRetryCount: retryCount,
+	}
+
+	event.LogTo(ctx, w.auditLogger, LevelAudit)
+}
+
+// LogStepFailed logs step failure.
+func (w *WorkflowAuditor) LogStepFailed(
+	ctx context.Context,
+	workflowID string,
+	stepID string,
+	duration time.Duration,
+	retryCount int,
+	_ error,
+) {
+	if !w.config.ShouldAuditEvent(EventTypeWorkflowStepFailed) {
+		return
+	}
+
+	source := w.extractSource(ctx)
+	subjects := w.extractSubjects(ctx)
+
+	event := NewAuditEvent(
+		EventTypeWorkflowStepFailed,
+		source,
+		OutcomeFailure,
+		subjects,
+		w.component,
+	)
+
+	target := map[string]string{
+		TargetKeyWorkflowID: workflowID,
+		TargetKeyStepID:     stepID,
+		TargetKeyType:       TargetTypeWorkflowStep,
+	}
+	event.WithTarget(target)
+
+	event.Metadata.Extra = map[string]any{
+		MetadataExtraKeyDuration:   duration.Milliseconds(),
+		MetadataExtraKeyRetryCount: retryCount,
+	}
+
+	event.LogTo(ctx, w.auditLogger, LevelAudit)
+}
+
+// LogStepSkipped logs conditional step skip.
+func (w *WorkflowAuditor) LogStepSkipped(
+	ctx context.Context,
+	workflowID string,
+	stepID string,
+	condition string,
+) {
+	if !w.config.ShouldAuditEvent(EventTypeWorkflowStepSkipped) {
+		return
+	}
+
+	source := w.extractSource(ctx)
+	subjects := w.extractSubjects(ctx)
+
+	event := NewAuditEvent(
+		EventTypeWorkflowStepSkipped,
+		source,
+		OutcomeSuccess,
+		subjects,
+		w.component,
+	)
+
+	target := map[string]string{
+		TargetKeyWorkflowID: workflowID,
+		TargetKeyStepID:     stepID,
+		TargetKeyType:       TargetTypeWorkflowStep,
+	}
+	event.WithTarget(target)
+
+	// Add condition as metadata
+	if condition != "" {
+		event.Metadata.Extra = map[string]any{
+			"condition": condition,
+		}
+	}
+
+	event.LogTo(ctx, w.auditLogger, LevelAudit)
+}
+
+// extractSource extracts source information from context.
+// For workflows, source is always local since they're internal orchestration.
+func (*WorkflowAuditor) extractSource(_ context.Context) EventSource {
```

Comment:
Follow-up idea: since vMCP uses both auditors (HTTP one logs the incoming call, this one logs the workflow internals), it'd be nice to have a correlation ID to tie them together.

Right now you can correlate by user + timestamp + tool name, but an explicit ID would make forensics easier when the same user fires off multiple requests. Maybe propagate `X-Request-ID` from the HTTP layer in a follow-up?

Reply:
https://github.com/stacklok/toolhive/issues/3032